### PR TITLE
[SPARK-5277][SQL] - SparkSqlSerializer doesn't always register user specified KryoRegistrators

### PR DESCRIPTION
**Pull request by mhseiden**
[SPARK-5277][SQL] - SparkSqlSerializer doesn't always register user specified KryoRegistrators

There were a few places where new SparkSqlSerializer instances were created with new, empty SparkConfs resulting in user specified registrators sometimes not getting initialized.

The fix is to try and pull a conf from the SparkEnv, and construct a new conf (that loads defaults) if one cannot be found.

The changes touched:
    1) SparkSqlSerializer's resource pool (this appears to fix the issue in the comment)
    2) execution.Exchange (for all of the partitioners)
    3) execution.Limit (for the HashPartitioner)

A few tests were added to ColumnTypeSuite, ensuring that a custom registrator and serde is initialized and used when in-memory columns are written.

### CODE REVIEW

**Comment by marmbrus** [comment]
ok to test

File: sql/core/src/main/scala/org/apache/spark/sql/execution/Exchange.scala [hunk 1]
```diff
@@ -46,6 +46,7 @@ case class Exchange(newPartitioning: Partitioning, child: SparkPlan) extends Una
     child.sqlContext.sparkContext.conf.getInt("spark.shuffle.sort.bypassMergeThreshold", 200)
 
   override def execute(): RDD[Row] = attachTree(this , "execute") {
+    lazy val sparkConf = Option(SparkEnv.get).map(_.conf).getOrElse(new SparkConf())
```
Comment (marmbrus) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/Exchange.scala]: All nodes should have a `sqlContext` which should have a `sparkContext`.  Can we just get this from there instead?

File: sql/core/src/test/scala/org/apache/spark/sql/columnar/ColumnTypeSuite.scala [hunk 2]
```diff
@@ -229,3 +267,23 @@ class ColumnTypeSuite extends FunSuite with Logging {
...
+    val b = input.readLong()
+    CustomClass(a,b)
+  }
+}
+
+private[columnar] final class Registrator extends KryoRegistrator {
+  override def registerClasses(kryo: Kryo) {
+    kryo.register(classOf[CustomClass], CustomerSerializer)
+  }
+}
```
Comment (marmbrus) [inline sql/core/src/test/scala/org/apache/spark/sql/columnar/ColumnTypeSuite.scala]: new line at end of file

**Comment by marmbrus** [comment]
Thanks for working on this.  This LGTM with two minor comments.

**Comment by yhuai** [comment]
Seems `new SparkSqlSerializer(new SparkConf(false))` is also used in `org.apache.spark.sql.execution.Limit`. Can you fix that as well?

**Comment by mhseiden** [comment]
@yhuai - The Limit case is covered as well.

@marmbrus - Not a problem! My most recent PR addresses your comments. Is this a change that could be backported to 1.2.x / 1.3.x, or will it only live in master?

**Comment by marmbrus** [comment]
Manually fixed conflicts while merging with master.  Thanks!

**Pull request closed without merging**